modules/dists/BlockDist.chpl

```diff
@@ -1538,9 +1538,62 @@ where this.sparseLayoutType == unmanaged DefaultDist &&
   return true;
 }
 
+proc BlockArr.canDoOptimizedSwap(other) {
+  var domsMatch = true;
+  var domShapesMatch = true;
+
+  if this.dom != other.dom { // no need to check if this is true
```

Comment:
For the dynamic fast followers we just check `dom.dist.dsiEqualDMaps`. Why do we need to check the domain too in this case?

Reply:
Because a domain 0..99 and one 1..100 over the same distribution aren't guaranteed to have equal-sized chunks on all locales that support pointer-swapping.  So today, we make a swap that should be legal but slow illegal.

Reply:
Why isn't that a problem for fast-followers too?

Reply:
I don't know, maybe it is?  You're saying fast followers don't currently check that the local number of elements to be iterated over are the same for both arrays?

Reply:
Fast follower checks are delegated to distributions. And at the very least Block checks whether domains' `whole`s are the same.

There is also a TODO there to relax that rule, which is already relaxed for swaps with this implementation. At least to some extent. (Arrays can be swapped fast, if the domains don't match exactly but their shapes match as long as their distribution has the some bounding box as the domain itself)

https://github.com/chapel-lang/chapel/blob/master/modules/dists/BlockDist.chpl#L1154

Reply:
Just to follow on, I verified that fast followers don't fire for the same kind of 0..99 / 1..100 case mentioned above (not that I doubted Engin's analysis so much as because it wouldn't completely surprise me if we didn't have tests for it, and Elliot so rarely asks questions that don't point to an actual problem).